File: 2021/day04-2.ts
type Cell = { num: number; check: boolean };
type Board = Cell[][];

function solution(content: string) {
  const lines = content.split("\n");
  const numbers = lines[0].split(",").map(Number);
  const boards = parseBoards(lines.slice(2).join("\n").split("\n\n"));
  const [winningBoard, winningNumber] = play(boards, numbers);
  return (
    winningNumber *
    winningBoard
      .flat()
      .filter(({ check }) => !check)
      .map(({ num }) => num)
      .reduce((acc, cur) => acc + cur, 0)
  );
}

function isWinningLine(line: Cell[]) {
  return line.every(({ check }) => check);
}

function transpose(board: Board) {
  return board.map((_, index) => board.map((row) => row[index]));
}

function isWinner(board: Board) {
  return board.some(isWinningLine) || transpose(board).some(isWinningLine);
}

function play(boards: Board[], numbers: number[]): [Board, number] {
  let currentBoards = boards.slice();
  let winningBoard: Board = [];
  let winningNumber = 0;
  for (const anumber of numbers) {
    currentBoards = currentBoards.map((board) =>
      board.map((row) =>
        row.map(({ num, check }) => ({ num, check: check || num === anumber }))
      )
    );
    const wbs = currentBoards.filter(isWinner);
    if (wbs.length) {
      winningBoard = wbs.at(0)!;
      winningNumber = anumber;
    }
    currentBoards = currentBoards.filter((board) => !isWinner(board));
  }
  return [winningBoard, winningNumber];
}

function parseBoards(blocks: string[]) {
  return blocks.map((block) =>
    block.split("\n").map((line) =>
      line
        .split(" ")
        .filter((item) => item)
        .map((item) => ({ check: false, num: Number(item) }))
    )
  );
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2021/day04-2.hs

type Cell = (Int, Bool)
type Board = [[Cell]]

solution :: String -> Int
solution content = wn * (sum . map fst . filter (not . snd) . foldl (++) [] $ wb)
    where
        result = play (parseBoards . drop 2 . lines $ content) (map read . splitOn ',' . head . lines $ content)
        wb = fst result
        wn = snd result

isWinningLine :: [Cell] -> Bool
isWinningLine = all (\(_, check) -> check)

isWinner :: Board -> Bool
isWinner board = any isWinningLine board || any isWinningLine (transpose board)

transpose :: [[a]] -> [[a]]
transpose board = [ [((!! row) . (!! col) $ board) | col <- [0..size]] | row <- [0..size]]
    where
        size = length board -1

winner :: [Board] -> Board
winner [] = []
winner (x:xs) 
    | isWinner x = x
    | otherwise = winner xs

play :: [Board] -> [Int] -> (Board, Int)
play _ [] = ([], 0)
play bs ns = helper ([], 0) bs ns
    where
        helper (wb, wn) bs [] = (wb, wn)
        helper (wb, wn) bs (n:ns)
            | length winningBoard > 0 = helper (winningBoard, n) stillNotWinning ns
            | otherwise = helper (wb, wn) stillNotWinning ns
            where
                current = [
                    [[(num, check || num == n) | (num, check) <- row] | row <- board] | board <- bs]
                winningBoard = winner current
                stillNotWinning = filter (not . isWinner) current

parseBoards :: [String] -> [Board]
parseBoards xs = helper [[]] xs
    where
        helper :: [Board] -> [String] -> [Board]
        helper boards [] = boards
        helper boards ("":xs) = helper (boards ++ [[]]) xs
        helper boards (x:xs) = helper ((init boards) ++ [last boards ++ [[((read y)::Int, False) | y <- words x]]]) xs

splitOn :: Char -> String -> [String]
splitOn c xs = helper [""] c xs
  where
    helper o c "" = o
    helper o c (x : "")
      | x == c = o
      | otherwise = init o ++ [last o ++ [x]]
    helper o c (x : xs)
      | x == c = helper (o ++ [""]) c xs
      | otherwise = helper (init o ++ [last o ++ [x]]) c xs

main :: IO ()
main = do readFile "day04.dat" >>= print . solution
